index.html

```diff
@@ -2706,6 +2884,54 @@ <h2>
         <h2>
 Metadata Structure
         </h2>
+
+        <p>
+A <a href="metadata-structure">metadata structure</a> is a common and interoperable data structure used to
+communicate information alongside a <a>DID</a> and <a>DID document</a>, and is specifically
+used for the <a>DID resolution</a> function definition in this specification.
+Use of a <a href="metadata-structure">metadata structure</a> is intended to compliment the information available
+within a <a>DID</a> or <a>DID document</a>, and therefore is best suited to carrying
+metadata as an input or output to the function process.
+        </p>
+        
+        <p>
+A <a href="metadata-structure">metadata structure</a> consists of a set of properties.
+Each property has a single name and a single value.
+The name of the property MUST be a single string. 
+Property names MUST be unique within a given set. 
+Each property name MUST map unambiguously to a single value within the set.
+Property names MUST be compared by the exact byte value.
+The value of a property MUST be a single string.
+The value of a property MAY be an empty string.
+Properties with null or equivalent values MUST be treated as
+equivalent to omitted properties within a given set.
+Property values MAY be repeated under different property names within a given set.
+Property definitions MAY impose additional constraints on the property value, so long as
+those constraints allow for unambiguous representation as a single string.
+        </p>
+        
+        <p>
+Properties within a <a href="metadata-structure">metadata structure</a> MAY be represented by an
+implementation in a variety of ways in an underlying protocol or system.
+For example, an implementation could represent a <a href="metadata-structure">metadata structure</a> with
+a set of HTTP-style headers, a JSON object, or an in-memory data map.
+Implementations MUST NOT require callers of functions to know the details of the underlying representation
+of the <a href="metadata-structure">metadata structure</a> in order to call the function.
+        </p>
+
+        <p>
+Properties within a <a href="metadata-structure">metadata structure</a> MAY be transformed by an
+implementation in order to facilitate their transmission in an underlying protocol or implementation.
+Any transformations MUST be bidirectional and lossless. 
+Implementations MUST NOT expose these transformations to callers of the <a>DID resolution</a> function and
+MUST NOT require callers of the <a>DID resolution</a> function to know about the transformations in order to call the function.
+For example, an implementation could prepend the string <code>DID-Document-</code> to the names of all <a href="did-document-metadata-properties"></a>
```

Comment:
I'm having trouble groking this section, and wording like:

"MUST NOT require callers of the <a>DID resolution</a> function to know about the transformations in order to call the function."

I feel like this language is not necessary, so long as the response interface is defined... callers are required to understand the interface... right?

As a DID Method author, I define support for JSON / CBOR / JSON-LD... resolve returns a representation (singular)... we believe that the representation can be transformed in a lossless bi-directional manner if we believe that the contract interface and registry are all that is needed to support that....This feels like we are now saying that DID Method implementers are responsible for ensuring that resolve supports all 3... instead of relying on the registry and the WG to make that transparent.... using post resolution middleware... in other words, all DID Methods are now required to produce all representations?... 

I guess what I am trying to say is: "Any transformations MUST be bidirectional and lossless. "... I would like to see a proof that its possible for at least JSON /JSON-LD/CBOR, before I see normative text saying its a requirement... I'm not sure I believe its possible.... and if its not, then we will end with 3 non interoperable DID Representations... Yikes!

Reply:
So this is the metadata, not the document itself. The metadata is not in JSON/CBOR/etc. The metadata is a map of strings: so a Javascript hash object, a python dict, a java Map<String, String>, etc. 

The transformation in question here is to and from that data structure into whatever underlying format there needs to be. That's one of the reasons that I think it's important for the metadata structure to be a LIMITED data structure in terms of Map[String -> String] and not objects, arrays, numbers, or anything else. This is representable in any possible underlying system in a deterministic way. 

The reversible transforms basically just means that if your protocol needs to use "DID-Resolution-Content-Type" to represent a field, you can't have someone call you with "DID-Resolution-Content-Type", you have to let them call you with "content-type", and you need to return "content-type" in the resulting map. This needs to be a strict requirement, but I'm struggling how to make it normatively enforceable. Thus the "MUST NOT expose" language above as an attempt to get to that.

I think a number of people are getting lost on this point, especially that the metadata is intended to be completely independent of the underlying representation. Is there a way to rephrase this to make it more clear what's going on here?

Reply:
This sections feels to me like it's delving a little bit too deep into the internals of the `resolve` function. It seems out of scope of did-core.

> if your protocol needs to use "DID-Resolution-Content-Type" to represent a field, you can't have someone call you with "DID-Resolution-Content-Type", you have to let them call you with "content-type", and you need to return "content-type" in the resulting map.

Understood. However, I still question its pertinence to did-core. This belongs in a transport-specific did-resolution spec.

Reply:
I can see how it can be read that way, but that's not the intent. The intent isn't to describe how any mapping can happen, but instead put limits on what a mapping can do. The real requirement here is that if a property is named `foo-bar`, then I should be able to call this function with a property named `foo-bar` and get the expected behavior (defined by the property) regardless of how the implementation represents `foo-bar` under the hood. I'll put in an alternative phrasing for this requirement that might capture it better.